packages/astro/src/core/pipeline.ts

```diff
@@ -75,58 +32,19 @@ export class Pipeline {
 	 *
 	 * It throws an error if the page can't be rendered.
 	 */
-	async #tryRenderRoute(
-		renderContext: Readonly<RenderContext>,
-		env: Readonly<Environment>,
-		mod: Readonly<ComponentInstance> | undefined,
-		onRequest?: MiddlewareHandler
+	async renderRoute(
+		componentInstance: ComponentInstance | undefined
 	): Promise<Response> {
-		const apiContext = createAPIContext({
-			request: renderContext.request,
-			params: renderContext.params,
-			props: renderContext.props,
-			site: env.site,
-			adapterName: env.adapterName,
-			locales: renderContext.locales,
-			routingStrategy: renderContext.routing,
-			defaultLocale: renderContext.defaultLocale,
-		});
-
-		switch (renderContext.route.type) {
-			case 'page':
-			case 'fallback':
-			case 'redirect': {
-				if (onRequest) {
-					return await callMiddleware(onRequest, apiContext, () => {
-						return renderPage({
-							mod,
-							renderContext,
-							env,
-							cookies: apiContext.cookies,
-						});
-					});
-				} else {
-					return await renderPage({
-						mod,
-						renderContext,
-						env,
-						cookies: apiContext.cookies,
-					});
-				}
-			}
-			case 'endpoint': {
-				return await callEndpoint(mod as any as EndpointHandler, env, renderContext, onRequest);
-			}
-			default:
-				throw new Error(`Couldn't find route of type [${renderContext.route.type}]`);
-		}
-	}
-
-	/**
-	 * Store a function that will be called before starting the rendering phase.
-	 * @param fn
-	 */
-	onBeforeRenderRoute(fn: PipelineHookFunction) {
-		this.#hooks.before.push(fn);
+		Reflect.set(this.renderContext.request, Symbol.for('astro.routeData'), this.renderContext.route)
+		const { renderContext, environment } = this;
+		const { defaultLocale, locales, params, props, request, routing: routingStrategy } = renderContext;
+		const { adapterName, logger, site, serverLike } = environment;
+		const apiContext = createAPIContext({ adapterName, defaultLocale, locales, params, props, request, routingStrategy, site });
+		const terminalNext = renderContext.route.type === 'endpoint'
```

Comment:
"terminal"? I can't understand the context of this word here

Reply:
It's the last `next()` that renders the matched route. As opposed to "non-terminal" `next()`s, which call the next middleware.

Renamed to `lastNext`.